adapters/http-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -1138,28 +1145,22 @@ protected boolean isCommandValid(final Command command, final Span currentSpan)
      * The created timer's ID is put to the routing context using key {@link #KEY_TIMER_ID}.
      *
      * @param ctx The device's currently executing HTTP request.
-     * @param responseReady The future to complete when the time has expired.
+     * @param responseReadyTracker The response ready tracker.
      * @param delaySecs The number of seconds to wait for a command.
-     * @param waitForCommandSpan The span tracking the command reception.
      */
     private void addCommandReceptionTimer(
             final RoutingContext ctx,
-            final AtomicBoolean requestProcessed,
-            final Handler<AsyncResult<Void>> responseReady,
-            final long delaySecs,
-            final Span waitForCommandSpan) {
+            final ResponseReadyTracker responseReadyTracker,
+            final long delaySecs) {
 
         final Long timerId = ctx.vertx().setTimer(delaySecs * 1000L, id -> {
 
             log.trace("time to wait [{}s] for command expired [timer id: {}]", delaySecs, id);
 
-            if (requestProcessed.compareAndSet(false, true)) {
+            if (responseReadyTracker.cancel(String.format("time to wait for command expired (%ds)", delaySecs))) {
                 // no command to be sent,
                 // send empty response
                 setTtdStatus(ctx, TtdStatus.EXPIRED);
-                waitForCommandSpan.log(String.format("time to wait for command expired (%ds)", delaySecs));
-                waitForCommandSpan.finish();
-                responseReady.handle(Future.succeededFuture());
```

Comment:
This change means that the execution order is now different.
`responseReadyTracker.cancel` already completes the `responseReady` promise, therefore with the new code `setTtdStatus(ctx, TtdStatus.EXPIRED);` will be run after that.

In most cases this will not be an issue since completing `responseReady` is followed by `closeCommandConsumer` which will do an I/O operation most of the times, so that execution will reach `setTtdStatus` before proceeding further along in `doUploadMessage` to make use of the updated ttd status.

But I could imagine a scenario where this is not the case (`closeCommandConsumer` returning directly with a failed Future), probably meaning that `setTtdStatus` is invoked too late then (i.e. after being used in `doUploadMessage`).

Therefore I think this code here should better keep the old execution order.